src/sentry/remote_subscriptions/consumers/result_consumer.py

```diff
@@ -79,20 +131,97 @@ def decode_payload(self, payload: KafkaPayload | FilteredPayload) -> T | None:
             )
         return None
 
+    def create_with_partitions(
+        self,
+        commit: Commit,
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        if self.parallel:
+            return self.create_thread_parallel_worker(commit)
+        else:
+            return self.create_serial_worker(commit)
+
+    def create_serial_worker(self, commit: Commit) -> ProcessingStrategy[KafkaPayload]:
+        return RunTask(
+            function=self.process_single,
+            next_step=CommitOffsets(commit),
+        )
+
     def process_single(self, message: Message[KafkaPayload | FilteredPayload]):
         result = self.decode_payload(message.payload)
         if result is not None:
             self.result_processor(result)
 
-    def create_serial_worker(self, commit: Commit) -> ProcessingStrategy[KafkaPayload]:
-        return RunTask(
-            function=self.process_single,
+    def create_thread_parallel_worker(self, commit: Commit) -> ProcessingStrategy[KafkaPayload]:
+        assert self.parallel_executor is not None
+        batch_processor = RunTask(
+            function=self.process_batch,
             next_step=CommitOffsets(commit),
         )
+        return BatchStep(
+            max_batch_size=self.max_batch_size,
+            max_batch_time=self.max_batch_time,
+            next_step=batch_processor,
+        )
 
-    def create_with_partitions(
-        self,
-        commit: Commit,
-        partitions: Mapping[Partition, int],
-    ) -> ProcessingStrategy[KafkaPayload]:
-        return self.create_serial_worker(commit)
+    def partition_message_batch(
+        self, message: Message[ValuesBatch[KafkaPayload]]
+    ) -> Generator[list[T]]:
+        """
+        Takes a batch of messages and partitions them based on the `build_payload_grouping_key` method.
+        Returns a generator that yields each partitioned list of messages.
+        """
+        batch = message.payload
+
+        batch_mapping: Mapping[str, list[T]] = defaultdict(list)
+        for item in batch:
+            assert isinstance(item, BrokerValue)
+
+            result = self.decode_payload(item.payload)
+            if result is None:
+                continue
+
+            key = self.build_payload_grouping_key(result)
+            batch_mapping[key].append(result)
+
+        # Number of messages that are being processed in this batch
+        metrics.gauge(
+            "remote_subscriptions.result_consumer.parallel_batch_count",
+            len(batch),
+            tags={"identifier": self.identifier, "mode": self.mode},
+        )
+        # Number of groups we've collected to be processed in parallel
+        metrics.gauge(
+            "remote_subscriptions.result_consumer.parallel_batch_groups",
+            len(batch_mapping),
+            tags={"identifier": self.identifier, "mode": self.mode},
+        )
+
+        yield from batch_mapping.values()
```

Comment:
Does this actually need to be a generator if it just yields one thing? or maybe I'm misunderstanding `yield from`?

Reply:
It returns a list of lists - so `yield from` will yield a list each time, one per unique group

Reply:
why does it need to be a generator though, why not just return the list[list[T]]

Reply:
Yeah, we could just return that... I don't remember why I did it this way

Reply:
Might have been related to me tinkering around with multi-proc, not sure